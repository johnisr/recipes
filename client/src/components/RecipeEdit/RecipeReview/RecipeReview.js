import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Button, Icon } from 'semantic-ui-react';
import { postRecipe } from '../../../actions/actions';
import { RecipeDetail } from '../../RecipeDetail/RecipeDetail';

const multiStringToArray = str =>
  str
    .split('\n')
    .map(val => val.trim())
    .filter(val => val);

const arrayToNumber = arr => {
  if (!arr[0]) return 0;
  return arr.reduce((prev, curr) => {
    let total = prev;
    if (curr.seconds) total += +curr.seconds;
    if (curr.minutes) total += +curr.minutes * 60;
    if (curr.hours) total += +curr.hours * 60 * 60;
    if (curr.days) total += +curr.days * 60 * 60 * 24;
    return total;
  }, 0);
};

const formatRecipe = recipeValues => {
  const obj = JSON.parse(JSON.stringify(recipeValues));
  if (obj.notes) obj.notes = multiStringToArray(obj.notes);
  if (obj.category) obj.category = multiStringToArray(obj.category);
  if (obj.preparation) {
    obj.preparation.forEach((section, index) => {
      if (section) {
        obj.preparation[index].body = multiStringToArray(
          obj.preparation[index].body
        );
      }
    });
  }
  if (!obj.ingredients) {
    return null;
  }
  obj.ingredients.forEach((section, index) => {
    if (section) {
      obj.ingredients[index].body = multiStringToArray(
        obj.ingredients[index].body
      );
    }
  });
  if (!obj.cooking) {
    return null;
  }
  obj.cooking.forEach((section, index) => {
    if (section) {
      obj.cooking[index].body = multiStringToArray(obj.cooking[index].body);
    }
  });
  obj.cookingTime = arrayToNumber(obj.cookingTime);
  obj.preparationTime = arrayToNumber(obj.preparationTime);
  return obj;
};

export class RecipeReview extends Component {
  onSubmit = async () => {
    const { formValues, history, onCancel } = this.props;
    try {
      const recipe = this.format(formValues);
      if (!recipe) throw new Error('Invalid form values');
      await this.props.postRecipe(recipe);
      history.push('/dashboard');
    } catch (err) {
      onCancel();
    }
  };
  format = formValues => formatRecipe(formValues);
  render() {
    const { onCancel, formValues } = this.props;
    return (
      <div>
        <h2>Review</h2>
        {formValues && <RecipeDetail review recipe={this.format(formValues)} />}
        <Button onClick={() => onCancel()}>Back</Button>
        <Button positive icon floated="right" onClick={this.onSubmit}>
          Submit Recipe
          <Icon name="mail" />
        </Button>
      </div>
    );
  }
}

const mapStatetoProps = state => ({
  formValues: state.form.recipeForm.values,
});

const mapDispatchToProps = dispatch => ({
  postRecipe: recipe => dispatch(postRecipe(recipe)),
});

export default connect(mapStatetoProps, mapDispatchToProps)(
  withRouter(RecipeReview)
);
