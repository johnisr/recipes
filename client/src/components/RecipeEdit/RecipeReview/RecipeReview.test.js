import React from 'react';
import { shallow } from 'enzyme';
import { RecipeReview } from './RecipeReview';

import formValuesArray from '../../../tests/fixtures/formValues';

test('should render correctly', () => {
  const wrapper = shallow(
    <RecipeReview
      onCancel={jest.fn()}
      formValues={formValuesArray[0].request}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should call onCancel on button click', () => {
  const onCancelSpy = jest.fn();
  const wrapper = shallow(<RecipeReview onCancel={onCancelSpy} />);
  wrapper
    .find('Button')
    .at(0)
    .simulate('click');
  expect(onCancelSpy).toHaveBeenCalled();
});

describe('When calling onSubmit', () => {
  let history;
  let postRecipe;
  let onCancelSpy;
  beforeEach(() => {
    history = { push: jest.fn() };
    postRecipe = jest.fn();
    onCancelSpy = jest.fn();
  });

  test('should format formValues', () => {
    const wrapper = shallow(
      <RecipeReview
        formValues={formValuesArray[0].request}
        history={history}
        postRecipe={postRecipe}
        onCancel={onCancelSpy}
        match={{ params: {} }}
      />
    );
    wrapper
      .find('Button')
      .at(1)
      .simulate('click');
    expect(postRecipe).toHaveBeenLastCalledWith(formValuesArray[0].response);
  });

  // test('should call history.push to /dashboard', () => {
  //   const wrapper = shallow(
  //     <RecipeReview
  //       formValues={formValues}
  //       history={history}
  //       postRecipe={postRecipe}
  //       onCancel={onCancelSpy}
  //     />
  //   );
  //   wrapper
  //     .find('button')
  //     .at(1)
  //     .simulate('click');
  //   expect(history.push).toHaveBeenLastCalledWith('/dashboard');
  // });

  test('should call onCancel when formValues are not valid', () => {
    const wrapper = shallow(
      <RecipeReview
        formValues={{}}
        history={history}
        postRecipe={postRecipe}
        onCancel={onCancelSpy}
      />
    );
    wrapper
      .find('Button')
      .at(1)
      .simulate('click');
    expect(onCancelSpy).toHaveBeenCalled();
  });
});
