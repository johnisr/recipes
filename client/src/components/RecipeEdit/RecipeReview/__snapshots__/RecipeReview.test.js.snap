// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecipeReview
    formValues={
        Object {
            "category": "test
        another test",
            "cooking": Array [
              Object {
                "body": "saute all ingredients",
                "title": "main",
              },
            ],
            "cookingTime": Array [
              Object {
                "days": "1",
                "hours": "0",
                "minutes": "0",
                "seconds": "0",
              },
            ],
            "ingredients": Array [
              Object {
                "body": "2 eggs
        2 ham
        1 tomato",
                "title": "main",
              },
            ],
            "name": "Test Recipe",
            "notes": "Some notes
        About the test recipe",
            "preparationTime": Array [
              Object {
                "hours": "1",
                "minutes": "1",
                "seconds": "1",
              },
            ],
            "summary": "A test recipe",
          }
    }
    onCancel={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Review
</h2>,
        <RecipeDetail
          recipe={
                    Object {
                              "category": Array [
                                "test",
                                "another test",
                              ],
                              "cooking": Array [
                                Object {
                                  "body": Array [
                                    "saute all ingredients",
                                  ],
                                  "title": "main",
                                },
                              ],
                              "cookingTime": 86400,
                              "ingredients": Array [
                                Object {
                                  "body": Array [
                                    "2 eggs",
                                    "2 ham",
                                    "1 tomato",
                                  ],
                                  "title": "main",
                                },
                              ],
                              "name": "Test Recipe",
                              "notes": Array [
                                "Some notes",
                                "About the test recipe",
                              ],
                              "preparationTime": 3661,
                              "summary": "A test recipe",
                            }
          }
          review={true}
/>,
        <div>
          <h5>
                    Add an Image
          </h5>
          <input
                    accept="image/*"
                    onChange={[Function]}
                    type="file"
          />
</div>,
        <Button
          as="button"
          color="yellow"
          onClick={[Function]}
          role="button"
>
          Back
</Button>,
        <Button
          as="button"
          disabled={false}
          floated="right"
          icon={true}
          onClick={[Function]}
          positive={true}
          role="button"
>
          Submit Recipe
          <Icon
                    as="i"
                    name="mail"
          />
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Review",
        },
        "ref": null,
        "rendered": "Review",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "recipe": Object {
            "category": Array [
              "test",
              "another test",
            ],
            "cooking": Array [
              Object {
                "body": Array [
                  "saute all ingredients",
                ],
                "title": "main",
              },
            ],
            "cookingTime": 86400,
            "ingredients": Array [
              Object {
                "body": Array [
                  "2 eggs",
                  "2 ham",
                  "1 tomato",
                ],
                "title": "main",
              },
            ],
            "name": "Test Recipe",
            "notes": Array [
              "Some notes",
              "About the test recipe",
            ],
            "preparationTime": 3661,
            "summary": "A test recipe",
          },
          "review": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h5>
              Add an Image
</h5>,
            <input
              accept="image/*"
              onChange={[Function]}
              type="file"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add an Image",
            },
            "ref": null,
            "rendered": "Add an Image",
            "type": "h5",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "accept": "image/*",
              "onChange": [Function],
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": "Back",
          "color": "yellow",
          "onClick": [Function],
          "role": "button",
        },
        "ref": null,
        "rendered": "Back",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": Array [
            "Submit Recipe",
            <Icon
              as="i"
              name="mail"
/>,
          ],
          "disabled": false,
          "floated": "right",
          "icon": true,
          "onClick": [Function],
          "positive": true,
          "role": "button",
        },
        "ref": null,
        "rendered": Array [
          "Submit Recipe",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "i",
              "name": "mail",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Review
</h2>,
          <RecipeDetail
            recipe={
                        Object {
                                    "category": Array [
                                      "test",
                                      "another test",
                                    ],
                                    "cooking": Array [
                                      Object {
                                        "body": Array [
                                          "saute all ingredients",
                                        ],
                                        "title": "main",
                                      },
                                    ],
                                    "cookingTime": 86400,
                                    "ingredients": Array [
                                      Object {
                                        "body": Array [
                                          "2 eggs",
                                          "2 ham",
                                          "1 tomato",
                                        ],
                                        "title": "main",
                                      },
                                    ],
                                    "name": "Test Recipe",
                                    "notes": Array [
                                      "Some notes",
                                      "About the test recipe",
                                    ],
                                    "preparationTime": 3661,
                                    "summary": "A test recipe",
                                  }
            }
            review={true}
/>,
          <div>
            <h5>
                        Add an Image
            </h5>
            <input
                        accept="image/*"
                        onChange={[Function]}
                        type="file"
            />
</div>,
          <Button
            as="button"
            color="yellow"
            onClick={[Function]}
            role="button"
>
            Back
</Button>,
          <Button
            as="button"
            disabled={false}
            floated="right"
            icon={true}
            onClick={[Function]}
            positive={true}
            role="button"
>
            Submit Recipe
            <Icon
                        as="i"
                        name="mail"
            />
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Review",
          },
          "ref": null,
          "rendered": "Review",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "recipe": Object {
              "category": Array [
                "test",
                "another test",
              ],
              "cooking": Array [
                Object {
                  "body": Array [
                    "saute all ingredients",
                  ],
                  "title": "main",
                },
              ],
              "cookingTime": 86400,
              "ingredients": Array [
                Object {
                  "body": Array [
                    "2 eggs",
                    "2 ham",
                    "1 tomato",
                  ],
                  "title": "main",
                },
              ],
              "name": "Test Recipe",
              "notes": Array [
                "Some notes",
                "About the test recipe",
              ],
              "preparationTime": 3661,
              "summary": "A test recipe",
            },
            "review": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h5>
                Add an Image
</h5>,
              <input
                accept="image/*"
                onChange={[Function]}
                type="file"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add an Image",
              },
              "ref": null,
              "rendered": "Add an Image",
              "type": "h5",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "accept": "image/*",
                "onChange": [Function],
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": "Back",
            "color": "yellow",
            "onClick": [Function],
            "role": "button",
          },
          "ref": null,
          "rendered": "Back",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": Array [
              "Submit Recipe",
              <Icon
                as="i"
                name="mail"
/>,
            ],
            "disabled": false,
            "floated": "right",
            "icon": true,
            "onClick": [Function],
            "positive": true,
            "role": "button",
          },
          "ref": null,
          "rendered": Array [
            "Submit Recipe",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "name": "mail",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
